### ISSUE TITLE
Fontimage can't generate previews of bitmap fonts (.bdf,.pcf,.otb)

### ISSUE THREAD

**Issue by loki7990** [fontforge/fontforge#4450]
### Important
Mark with [x] to select. Leave as [ ] to unselect.

### When reporting a bug/issue:
- [ x] Screenshot
- [ x] The FontForge version and the operating system you're using
- [ x] The behavior you expect to see, and the actual behavior
- [ ] Steps to reproduce the behavior 
- [ ] \(optional) Possible solution/fix/workaround

### When you open an issue for a change/improvement/feature request:
- [ x] A description of the problem you're trying to solve, including _why_ you think this is a problem
- [ ] If the feature changes current behavior, reasons why your solution is better
- [ ] \(optional) Possible solution/fix/workaround

Hello,

According to the fontimage man page, `The program fontimage loads a font, which may be in any format fontforge(1) can read, and then produces an image showing representative glyphs of the font.` Fontforge is able to open and work with bitmap font types such as `.bdf`, `.pcf`, and `.otb` so I would expect that when I use the `fontimage` command I can preview these fonts in an image. Unfortunately, only `.otf` fonts are successfully previewed with the `fontimage` command. I believe this should be fixed because as someone who keeps many bitmap fonts locally, I'd like to quickly be able to preview the fonts rather than having to change program settings each and every time. This fix would greatly improve the usefulness of font image preview in programs such as ranger (see https://github.com/ranger/ranger/issues/2100). Below are screenshots to show what I'm seeing. The white image is an attempt to create a preview of a `.pcf` font while the successful one was a `.otf` font. I use `fontforge 20200314-3` on Arch Linux.
![2020-09-16_10-40-59-676x379](https://user-images.githubusercontent.com/57952969/93353692-358cc200-f80a-11ea-9914-e42b14be2c29.png)
![2020-09-16_10-41-05-871x259](https://user-images.githubusercontent.com/57952969/93353706-3887b280-f80a-11ea-8f9c-8d9ec8ef2d5a.png)

**Comment by ctrlcctrlv** [comment]
Hello

You need to realize that you're asking about something that hasn't been updated in seven years ([`fontimage.pe`](https://github.com/fontforge/fontforge/blob/master/fontforgeexe/fontimage.pe)), and which I didn't even know existed. I had to look at the manpage myself and see "&copy; 2006 George Williams" to even be sure that `fontimage` has something to do with FontForge. I thought it was maybe a third party script.

When we deprecated the native scripting system ([#D2](https://github.com/fontforge/fontforge/wiki/Community-guidelines#D2)), I don't know that we knew this existed. Maybe the others did, I don't recall taking this into consideration.

I think these three tools&mdash;`fontlint`, `fontimage` and `sfddiff`&mdash;need reimplementation in Python to make progress. The specific issue is that the functions which are being used as far as I know only work on outline fonts.

https://github.com/fontforge/fontforge/blob/c3468cbd0320c152c0cbf762b9e2b63642d9c65f/fontforgeexe/fontimage.pe#L99
https://github.com/fontforge/fontforge/blob/c3468cbd0320c152c0cbf762b9e2b63642d9c65f/fontforge/scripting.c#L2573
https://github.com/fontforge/fontforge/blob/c3468cbd0320c152c0cbf762b9e2b63642d9c65f/fontforge/sflayout.c#L1176

`sflayout.c` is particularly ancient, untouched since 2008.

I suppose if someone wants to fix this, we need to figure out why `LI_FDDrawChar` isn't drawing bitmaps. 

https://github.com/fontforge/fontforge/blob/c3468cbd0320c152c0cbf762b9e2b63642d9c65f/fontforge/sflayout.c#L73

**Comment by loki7990** [comment]
Hello,

I had a feeling regarding that 2006 at the end... A small hope I had was that it was a typo/ simply forgotten to be updated. Unfortunately I don't have the time to implement this, so I'm ok with closing this and simply removing those tools if they aren't going to receive upkeep. Since you mentioned `hb-view` over in the ranger issue, it makes more sense to use that moving forward.

**Comment by ctrlcctrlv** [comment]
Yes, for your specific issue `hb-view` makes the most sense, because even if this gets fixed, it's still going to be using FontForge's rasterizer which doesn't support all of the OpenType magic that HarfBuzz does.

I think these tools should be removed from the default distribution as I don't see any particular reason to use `fontimage` over `hb-view` in any situation. `fontlint` is much better handled by other tools as well. Only one that -might- still be useful is `sfddiff`, but I've never used it. What do you think @jtanx @frank-trampe ? Users who really need to use them with new versions can pull them out of old Git commits and run them via `-lang=ff -script ...`. No?

**Comment by skef** [comment]
I use `sfddiff` quite often, and while it's somewhat superseded by `ttx` it isn't entirely so, particularly when it comes to questions about FontForge's own font-level export code.

I've never touched the other two but that doesn't mean much.

**Comment by ctrlcctrlv** [comment]
And I mean @loki7990 , I have a pretty good idea of what people use FontForge for nowadays, and I very much doubt we're ever going to receive a PR for either of the three tools. I was surprised to learn they existed...I have used FontForge since George was still around. That doesn't bode well. :-) 

I can't afford to sink time into effectively reimplementing `hb-view`, and @behdad is way better at this stuff than me anyway so I wouldn't want to. I think we should just write to the mailing list that we recommend:

* Users of `fontimage` use `hb-view`;
* Users of `fontlint` use `fontbakery` (https://github.com/googlefonts/fontbakery);
* Users of `sfddiff` use the regular `diff` (or `git --diff`) in combination with `sfdnormalize` (https://pypi.org/project/sfdnormalize/).

@skef does `sfddiff` do something other than diff SFD's?

**Comment by skef** [comment]
`sfddiff` accepts font files in the same way that FontForge does, so you can directly pass two `.otf` files (for example) on the command line and the tool will present the differences. (The name is a slight misnomer in that what's being compared are the FontForge-internal representations that correspond to SFD content rather than literal SFD files.)

**Comment by ctrlcctrlv** [comment]
I see. The bulk of the operation of `sfddiff` is done by `CompareFonts`:

https://github.com/fontforge/fontforge/blob/c3468cbd0320c152c0cbf762b9e2b63642d9c65f/fontforge/scripting.c#L8362

Mercifully, a Python version already seems to exist:

https://github.com/fontforge/fontforge/blob/c3468cbd0320c152c0cbf762b9e2b63642d9c65f/fontforge/python.c#L14496

So...the most useful one is the easiest to replace with a Python version.

**Comment by skef** [comment]
I guess, but couldn't we also just leave `sfddiff` regardless of what we do with the other tools?

**Comment by ctrlcctrlv** [comment]
Certainly :-)

**Comment by khaledhosny** [comment]
All the three tools have been implemented in C for a long long time, the scripts are a very thin wrappers.

**Comment by ctrlcctrlv** [comment]
Good news :)

Mostly I don't think it's a good idea to advertise features that have no practical application; I see fontimage as such a feature in a world with `hb-view`. And fontlint likewise in a world with `fontbakery`.

I see the application of `sfddiff`...for developers.

**Comment by khaledhosny** [comment]
`fontimage` if garbage, I agree (though IIRC, it has some code to decide what character sample to use, `hb-view` just renders whatever text you give it). I have used `fontlint` quite often and it catches real issues, while `fontbakery` is a very opinionated Google tool that lumps technical, stylistic, and bogus requirements together and I always have to double check whether the issues it flags are real or not (often than not, they aren’t).

**Comment by ctrlcctrlv** [comment]
I see, thanks for that perspective. I took `fontbakery` to be quite gospel in terms of what are best standards. Perhaps it would be helpful if you could compile a list of the Fontbakery codes that are just Google's opinion, as unfortunately `fontlint` isn't that useful for variable fonts so Fontbakery is the only option. Of course, only should the idea interest you. :-)
